connaisseur/validators/cosign/cosign_validator.py

```diff
@@ -21,44 +22,110 @@
-            )
-        except StopIteration as err:
-            msg = 'Trust root "{key_name}" not configured for validator "{validator_name}".'
-            raise NotFoundException(
-                message=msg, key_name=key_name, validator_name=self.name
-            ) from err
-        return Key("".join(key))
-
     async def validate(
         self, image: Image, trust_root: str = None, **kwargs
     ):  # pylint: disable=arguments-differ
-        key = self.__get_key(trust_root)
-        return self.__get_cosign_validated_digests(image, key).pop()
 
-    def __get_cosign_validated_digests(self, image: Image, key: Key):
+        required = kwargs.get("required", [])
+        # if not configured, `threshold` is 1 if trust root is not "*" or
+        # `required` is specified and number of trust roots otherwise
+        threshold = kwargs.get(
+            "threshold",
+            (int(any(required)) or len(self.trust_roots)) if trust_root == "*" else 1,
+        )
+
+        self.vals = self.__get_pinned_keys(trust_root, required)
+
+        tasks = [self.__validation_task(k, image) for k in self.vals.keys()]
+        await asyncio.gather(*tasks)
+
+        return CosignValidator.__apply_policy(
+            vals=self.vals, threshold=threshold, required=required
+        )
+
+    def __get_pinned_keys(self, key_name: str = None, required: list = None):
+        """
+        Extract the pinned key(s) selected for validation from the list of trust roots.
+        """
+        key_name = key_name or "default"
+
+        if key_name == "*":
+            keys = {
+                k["name"]: {
+                    "name": k["name"],
+                    "key": Key("".join(k["key"])),
+                    "digest": None,
+                    "error": None,
+                }
+                for k in self.trust_roots
+            }
+
+            # check if all `required` are present in trust roots
+            not_found = set(required) - {k["name"] for k in self.trust_roots}
+            if not_found:
+                msg = 'Trust roots "{key_names}" not configured for validator "{validator_name}".'
+                raise NotFoundException(
+                    message=msg,
+                    key_names=", ".join(not_found),
+                    validator_name=self.name,
+                )
+        else:
+            try:
+                key = next(k["key"] for k in self.trust_roots if k["name"] == key_name)
+            except StopIteration as err:
+                msg = 'Trust root "{key_name}" not configured for validator "{validator_name}".'
+                raise NotFoundException(
+                    message=msg, key_name=key_name, validator_name=self.name
+                ) from err
+            keys = {
+                key_name: {
+                    "name": key_name,
+                    "key": Key("".join(key)),
+                    "digest": None,
+                    "error": None,
+                }
+            }
```

Comment:
Code duplication from above. Suggesting to construct `keys` only once (via iteration), and do the special-case handling when preparing the iteration input:

```python
trust_roots = self.trust_roots
if key_name != "*":
    trust_roots = filter(lambda key: key["name"] = key_name, trust_roots)
if not trust_roots:
    msg = "..."
    raise ...

return {
    ...
    for key in trust_roots
}
```

Reply:
took a bit of different path...let me know what you think